File: examples/week02/factory/j/factory/ShapeFactory.java
package j.factory;

public class ShapeFactory {
    public Shape newShape(String shapeType) {
        if (shapeType == null)
            return null;
        if (shapeType.equalsIgnoreCase("rectangle"))
            return new Rectangle();
        else if (shapeType.equalsIgnoreCase("square"))
            return new Square();
        else if (shapeType.equalsIgnoreCase("circle"))
            return new Circle();
        else
            return null;
    }
}

File: examples/week02/factory/s/factory/ShapeFactory.scala
package s.factory

trait Shape {
  def draw: Unit
}

class Rectangle extends Shape {
  override def draw: Unit = println("drawing Rectangle!")
}

class Square extends Shape {
  override def draw: Unit = println("drawing Square")
}

class Circle extends Shape {
  override def draw: Unit = println("drawing Circle")
}

object ShapeFactory {
  def apply(shapeType: String): Option[Shape] =
    if (shapeType.toLowerCase == "rectangle") Some(new Rectangle)
    else if (shapeType.toLowerCase == "square") Some(new Square)
    else if (shapeType.toLowerCase == "circle") Some(new Circle)
    else None
}
